larpandoracontent/LArReclustering/ThreeDReclusteringAlgorithm.cc

```diff
@@ -0,0 +1,483 @@
+{
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+StatusCode ThreeDReclusteringAlgorithm::Run()
+{
+    //Only proceed if successfuly get shower pfo list, and it has at least one pfo
+    const PfoList *pShowerPfoList(nullptr);
+    if (!((STATUS_CODE_SUCCESS == PandoraContentApi::GetList(*this, m_pfoListName, pShowerPfoList)) && pShowerPfoList->size()))
+    {
+        return STATUS_CODE_SUCCESS;
+    }
+
+    m_PfosForReclusteringListName = "newShowerParticles3D";
+    PfoList unchangedPfoList;   
+
+    //Save current pfo list name, so that this can be set as current again at the end of reclustering
+    std::string initialPfoListName;
+    PANDORA_RETURN_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::GetCurrentListName<Pfo>(*this, initialPfoListName));
+
+    //Some pfos are shower-like and yet include track-like 3D clusters. For the moment I don't want to deal with these.
+    const ClusterList *pShowerClusters(nullptr);
+    PandoraContentApi::GetList(*this, "ShowerClusters3D", pShowerClusters);
+
+    if(!pShowerClusters)
+        return STATUS_CODE_NOT_FOUND;
+
+    for (const Pfo *const pShowerPfo : *pShowerPfoList)
+    {
+        ClusterList clusterList3D;
+        LArPfoHelper::GetThreeDClusterList(pShowerPfo, clusterList3D);
+       
+        if(!this->PassesCutsForReclustering(pShowerPfo))
+        {
+            unchangedPfoList.push_back(pShowerPfo);
+            continue;
+        }
+
+        //Some pfos are shower-like and yet include track-like 3D clusters. For the moment I don't want to deal with these.
+        if(pShowerClusters->end() == std::find(pShowerClusters->begin(), pShowerClusters->end(), clusterList3D.front()))
+            continue;
+
+        CaloHitList caloHitList3D;
+        clusterList3D.front()->GetOrderedCaloHitList().FillCaloHitList(caloHitList3D);
+
+		//Create a variable for the minimum figure of merit and initialize to initial FOM
+        float minimumFigureOfMerit(this->GetFigureOfMerit(caloHitList3D));
+
+        //Free the hits in this cluster, so that they are not owned by the original pfo, and are available for reclustering
+        PANDORA_THROW_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::RemoveFromPfo(*this, pShowerPfo, clusterList3D.front()));
+
+        // Initialize reclustering with these local lists
+        std::string currentClustersListName;
+        PANDORA_RETURN_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::GetCurrentListName<Cluster>(*this, currentClustersListName));
+        PANDORA_RETURN_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::ReplaceCurrentList<Cluster>(*this, "ShowerClusters3D"));
+
+		//Split the calo hit list into a new set of calo hit lists, taking the best outcome out of different algorithms
+        std::vector<CaloHitList*> newCaloHitListsVector, minimumFigureOfMeritCaloHitListsVector, initialCaloHitListsVector;
```

Comment:
Should this be a pointer? - The answer might be yes here, because we're getting into Pandora list management territory, but flagging it here in case it's not necessary.